tests/test_check_parallel.py

```diff
@@ -0,0 +1,387 @@
+# For details: https://github.com/PyCQA/pylint/blob/master/COPYING
+
+# pylint: disable=protected-access,missing-function-docstring,no-self-use
+
+import collections
+import os
+
+import pytest
+
+import pylint.interfaces
+import pylint.lint.parallel
+from pylint.checkers.base_checker import BaseChecker
+from pylint.lint import PyLinter
+from pylint.lint.parallel import _worker_check_single_file as worker_check_single_file
+from pylint.lint.parallel import _worker_initialize as worker_initialize
+from pylint.lint.parallel import check_parallel
+from pylint.testutils import TestReporter as Reporter
+
+
+def _gen_file_data(idx=0):
+    """ Generates a file to use as a stream """
+    filepath = os.path.abspath(
+        os.path.join(os.path.dirname(__file__), "input", "similar1")
+    )
+    file_data = (
+        "--test-file_data-name-%d--" % idx,
+        filepath,
+        "--test-file_data-modname-%d--" % idx,
+    )
+    return file_data
+
+
+def _gen_file_datas(count=1):
+    file_infos = [_gen_file_data(idx) for idx in range(count)]
+    return file_infos
+
+
+class SequentialTestChecker(BaseChecker):
+    """ A checker that does not need to consolidate data across run invocations """
+
+    __implements__ = (pylint.interfaces.IRawChecker,)
+
+    name = "sequential-checker"
+    test_data = "sequential"
+    msgs = {"R9999": ("Test", "sequential-test-check", "Some helpful text.",)}
+
+    def __init__(self, linter, *args, **kwargs):
+        super().__init__(linter, *args, **kwargs)
+        self.data = []
+        self.linter = linter
+
+    def process_module(self, _astroid):
+        """ Called once per stream/file/astroid object """
+        # record the number of invocations with the data object
+        record = self.test_data + str(len(self.data))
+        self.data.append(record)
+
+
+class ExtraSequentialTestChecker(SequentialTestChecker):
+    """ A checker that does not need to consolidate data across run invocations """
+
+    name = "extra-sequential-checker"
+    test_data = "extra-sequential"
+
+
+class ThirdSequentialTestChecker(SequentialTestChecker):
+    """ A checker that does not need to consolidate data across run invocations """
+
+    name = "third-sequential-checker"
+    test_data = "third-sequential"
+
+
+class TestCheckParallelFramework:
+    """ Tests the check_parallel() function's framework """
+
+    def setUp(self):
+        self._prev_global_linter = pylint.lint.parallel._worker_linter
+
+    def tearDown(self):
+        pylint.lint.parallel._worker_linter = self._prev_global_linter
+
+    def test_worker_initialize(self):
+        linter = PyLinter(reporter=Reporter())
+        worker_initialize(linter=linter)
+        assert pylint.lint.parallel._worker_linter == linter
+
+    def test_worker_check_single_file_uninitialised(self):
+        pylint.lint.parallel._worker_linter = None
+        with pytest.raises(AttributeError):
```

Comment:
It's not obvious what `AttributeError` this is and why it has to be raised. I see two possible solutions here, maybe complementary:
* Use the context manager to verify the exception message with `pytest.raises(AttributeError, msg)`
* Change the `AtttributeError` to a user-defined expection that gets raised when the condition is not met.

Reply:
You're completely right, and normally I'd agree and add some custom explicit/beautiful expception instead. This just shows a possible error state that the parallel system isn't handling properly yet, but captures the current behavior and there will help identify regressions, not much more that that really - but it is still valuable.

Don't forget the intent here is to capture the current state so the parallel system can be refoactored at pace, and made _significantly_ faster, so most of the code is likely to change.

I've added a msg.